src/main/java/seedu/address/model/ContactTagIntegrationManager.java

```diff
@@ -0,0 +1,188 @@
+package seedu.address.model;
+
+import java.util.HashSet;
+import java.util.Set;
+import java.util.function.Consumer;
+
+import seedu.address.model.person.Person;
+import seedu.address.model.tag.ReadOnlyTagTree;
+import seedu.address.model.tag.Tag;
+import seedu.address.model.tag.TagTree;
+import seedu.address.model.tag.TagTreeImpl;
+
+/**
+ * Class that controls {@code Tag} relationships and interactions between {@code Tag}s and {@Code Person}s.
+ */
+public class ContactTagIntegrationManager {
+
+    private AddressBook addressBook;
+    private TagTree tagTree;
+
+    /**
+     * Creates a ContactTagIntegrationManager from the given {@code addressBook} and a new tagTree.
+     */
+    public ContactTagIntegrationManager(AddressBook addressBook) {
+        this.addressBook = addressBook;
+        tagTree = new TagTreeImpl();
+    }
+
+    /**
+     * Creates a ContactTagIntegrationManager from the given {@code addressBook} and {@code tagTree}.
+     */
+    public ContactTagIntegrationManager(AddressBook addressBook, ReadOnlyTagTree tagTree) {
+        this(addressBook);
+        this.tagTree.copy(tagTree);
+    }
+
+    public AddressBook getAddressBook() {
+        return addressBook;
+    }
+
+    public TagTree getTagTree() {
+        return tagTree;
+    }
+
+    /**
+     * Returns a {@code set} or {@code person}s that falls under the given {@code tag} and all its sub-tags.
+     *
+     * @param tag {@code tag} to begin search for matching {@code person}s.
+     * @return Set of Persons falling under either the argument {@code tag} or its sub-tags.
+     */
+    public Set<Person> getAllPersonsUnderTag(Tag tag) {
+        Set<Person> finalSet = new HashSet<>();
+        Consumer<Tag> consumer = new Consumer<Tag>() {
+            @Override
+            public void accept(Tag tag) {
+                if (addressBook.getPersonsWithTag(tag) == null) {
+                    return;
+                }
+                finalSet.addAll(addressBook.getPersonsWithTag(tag));
+                tagTree.getSubTagsOf(tag).stream().forEach(subTag -> accept(subTag));
+            }
+        };
+        consumer.accept(tag);
+        return finalSet;
+    }
+
+    /**
+     * Edits the sub-tags of a given {@code supertag}.
+     * If a sub-tag to add is already present, no change will be performed.
+     * If a sub-tag to be removed is not present, no change will be performed.
```

Comment:
perhaps both these sentences could be combined into one.

Reply:
I think the individual behaviors have to be explicit. In this case, the upcoming `editTagCommand` will specify tags to add and tags to remove as two separate tag sets. As such, I wanted to be clear here on the behavior of the method for each case.

Reply:
gotcha.